server_src/services/octoprint/octoprint-api-client.service.js

```diff
@@ -0,0 +1,127 @@
+const apiPrinterProfiles = apiBase + "/printerprofiles";
+const apiSystem = apiBase + "/system";
+const apiSystemInfo = apiSystem + "/info";
+const apiSystemCommands = apiSystem + "/commands";
+const apiUsers = apiBase + "/users";
+const apiLogin = (passive = true) =>
+  apiBase + "/login" + (passive ? "?passive=true" : "");
+
+const apiPluginManager = apiBase + "/plugin/pluginmanager";
+const apiPluginManagerRepository1_6_0 = octoPrintBase + "plugin/pluginmanager/repository";
+const apiSoftwareUpdateCheck = (force) =>
+    octoPrintBase + "plugin/softwareupdate/check" + (force ? "?force=true" : "");
+const apiPluginPiSupport = apiBase + "/plugin/pi_support";
+
+const printerValidationErrorMessage = "printer apiKey or URL undefined";
+
+class OctoprintApiClientService extends OctoprintApiService {
+  constructor(timeoutSettings) {
+    super(timeoutSettings);
+  }
+
+  static validatePrinter(printer) {
+    if (!printer.apikey || !printer.printerURL) {
+      throw printerValidationErrorMessage;
+    }
+  }
+
+  async postPrinter(printer, route, data, timeout = false) {
+    OctoprintApiClientService.validatePrinter(printer);
+    return super.post(printer.printerURL, printer.apikey, route, data, timeout);
+  }
+
+  async getWithOptionalRetry(printer, route, retry = false) {
+    OctoprintApiClientService.validatePrinter(printer);
+    if (retry) {
+      return await this.getRetry(printer.printerURL, printer.apikey, route);
+      // .then(r => r.json());
+    } else {
+      return await this.get(printer.printerURL, printer.apikey, route);
+      // .then(r => r.json());
+    }
+  }
+
+  async getSettings(printer, retry = false) {
+    return this.getWithOptionalRetry(printer, apiSettingsPart, retry);
+  }
+
+  /**
+   * List files recursively or not.
+   * @param printer
+   * @param recursive
+   * @param retry
+   * @returns {Promise<*|Promise|Promise<unknown> extends PromiseLike<infer U> ? U : (Promise|Promise<unknown>)|*|undefined>}
+   */
+  async getFiles(printer, recursive = false, retry = false) {
+    return this.getWithOptionalRetry(printer, apiFiles(recursive), retry);
+  }
+
+  /**
+   * Get a specific file
+   * @param printer
+   * @param path
+   * @param retry
+   * @returns {Promise<*|Promise|Promise<unknown> extends PromiseLike<infer U> ? U : (Promise|Promise<unknown>)|*|undefined>}
+   */
+  async getFile(printer, path, retry = false) {
+    return this.getWithOptionalRetry(printer, apiFile(path), retry);
+  }
+
+  async getConnection(printer, retry = false) {
+    return this.getWithOptionalRetry(printer, apiConnection, retry);
+  }
+
+  async getPrinterProfiles(printer, retry = false) {
+    return this.getWithOptionalRetry(printer, apiPrinterProfiles, retry);
+  }
+
+  async getPluginManager(printer, retry = false) {
+    const printerManagerApiCompatible = checkPluginManagerAPIDeprecation(printer.octoPrintVersion);
+
+    const route = printerManagerApiCompatible ? apiPluginManagerRepository1_6_0 : apiPluginManager;
```

Comment:
I love this kind of coding, really clean short functions with seasoning in the right places